### PR TITLE
src/mame/etc/gen_device_defs.py: major overhaul of device file generation:

### PR DESCRIPTION
**Pull request by angelosa**
* Converted template_device.* to use basic Jinja2-like patterns;
* Add support for creating device defs in the known MAME folders where a device file is expected;
* Add license field validation;
* Improve documentation;

### CODE REVIEW

**Review by cuavas — COMMENTED** [review]
Well besides the mess this makes of the templates (member function with 8-bit data and 32-bit memory mask, inconsistent use of standard and abbreviated integer types, etc.), how much value does this really provide?

Creating a device class isn’t difficult.  Putting it behind a script obfuscates the process somewhat, and makes people less aware of what is and isn’t needed (for example `device_add_mconfig` and `device_reset` aren’t pure virtual members).  It also ties the shortname and class name together which isn’t desirable – the allowable character sets are different, length limits are different, classes can be namespaced while short names need a prefix.

To me, this seems to occupy an uncomfortable middle ground.  I can see the potential attraction of an all-singing, all-dancing GUI wizard interface with checkboxes for the optional methods to override, that generates empty I/O port and ROM definitions if you ask for them, can generate skeleton implementations of common interfaces, etc.  But this is just a glorified search/replace.  The extra effort to ensure it doesn’t rot is probably more than the effort it will save anyone.

**Comment by angelosa** [comment]
So what’s your concrete solution in resolving my headaches when needing to create a new device in MAME? Need to unpleasantly touch 4-5 files scattered across the tree, in two different languages. And MAME domain language is pretty much a damn jigsaw, especially for people not accustomed with this.

PS: if I search of “MAME Add New Device” [this](https://wiki.mamedev.org/index.php/MAME_Device_Basics) is very unhelpful, speaking of rot …

**Comment by angelosa** [comment]
Code rot can be suppressed by adding a CI pipeline specifically for this. 
For the encoder itself this, but it needs a parser engine and afaict the most popular is Jinja2 ([C++](https://github.com/jinja2cpp/Jinja2Cpp), [Python](https://pypi.org/project/Jinja2/), [relevant SO discussion](https://stackoverflow.com/questions/2966618/c-code-generation-with-python)). 
And of course this also implies that we need a virtual env setup for Python, which needs a separate discussion. 

Consider this a PoC, wanna see the C++ MAME alternative? [Here](https://github.com/mamedev/mame/blob/master/src/lib/netlist/prg/nltool.cpp):
- Copyright holder is fixed with Couriersud as author, while I respect the work on netlist here (which is also not my domain) 1356 lines looks a bit too much for classifying it as a "simple" tool lol;
- There are Russian strings at the very end under `#if 0`;

**Comment by cuavas** [comment]
>So what’s your concrete solution in resolving my headaches when needing to create a new device in MAME? Need to unpleasantly touch 4-5 files scattered across the tree, in two different languages.

Writing an empty device class is trivial – the only things you need are a public constructor with the standard signature and a `device_start()` member function.  The device declaration and definition macros are trivial and hide the details of how it actually creates the device type object.  Are you seriously trying to say that’s an inordinate amount of effort to type?

You will (or at least should) spend a lot more time actually designing the device class than typing the minimal amount of boilerplate to create an empty device.

This doesn’t make it any more convenient to add a device to the build scripts.  You still need to choose a name for the inclusion variable, set the headers to trigger inclusion on, and add it in an appropriate spot.  Trying to automate that will require a proper parser for Lua (or CMake script once we go there), and a bunch more prompts.  No, a half-arsed parser that requires stuff to be formatted precisely how it expects won’t be acceptable.

>And MAME domain language is pretty much a damn jigsaw, especially for people not accustomed with this.

We’re aiming to make MAME use idiomatic C++ so that it *isn’t* a puzzle to a competent C++ developer, and we’ve done pretty well so far.  We’ve gotten rid of most of the MAME metalanguage:
* Device classes are plain C++ classes with only a single pure virtual member function that needs to be implemented.
* Machine configuration is plain C++, it no longer relies on a fragile, opaque mess of hidden variables, macros dynamic casts.
* Address maps are also plain C++ – you can find out what the valid operations on the `map` argument are just by looking at the class declaration, you don’t need to understand a pile of magical macros.
* You no longer need to write a pile of macros to make it possible to use a device (cue horrifying flashback to the `MCFG_` mess).

There are three major remaining pieces of metalanguage (ignoring minor ones like the bits and pieces associated with media image formats):
* ROM definitions.
* I/O port definitions.
* System definition macros.

Your empty device class generator doesn’t help with any of these, and it wouldn’t be practical to try to generate that kind of stuff without some kind of GUI for configuring it.

Your generator doesn’t help with any of the repetitive parts of writing a device either, like providing configuration methods for devcb/devdelegate, object finders, etc.  And once again, this isn’t going to be something you’d be able to do practically in a command-line python script.

In purely practical terms, you *really* need a wizard-style GUI for configuring this kind of thing, trying to get the developer to name all their members and methods in a text-based interface is impractical and error-prone.  Once stuff starts to get complex, you need a real code generation framework.  A report generation library isn’t up to the task.  Trying to do it with a report generation library or template system always ends up producing terrible-looking code.

More philosophically, you often don’t know what all the required integration points are going to be when you first create your device class.  You probably don’t know everything you need to know when first creating the class.  If you want to be able to *add* integration points to a device class, then you’re getting well into the territory of refactoring tools.  At this point, you need a solid C++ parser as well as a generator, you need a bunch of heuristics to guess where to place the additions, and you need some kind of UI for letting the developer adjust that.

>PS: if I search of “MAME Add New Device” [this](https://wiki.mamedev.org/index.php/MAME_Device_Basics) is very unhelpful, speaking of rot …

We all know the documentation on the wiki was a failure.  It never had a proper table of contents so it was hard to find things.  It was never updated along with the code, even in cases where the people who wrote the wiki pages were the ones changing the code.  This is an example of why I favour Doxygen-style API documentation – at least by keeping it with the relevant code, it can’t rot due to being out of sight, out of mind.

But putting aside the fact that the wiki isn’t in great shape at the moment due to PHP version issues, how much have you actually contributed to helping keep the documentation up-to-date and relevant?  It’s easy to take cheap shots, but if *everyone* maintains this attitude of “documentation is something other people do”, rot is inevitable.

>Code rot can be suppressed by adding a CI pipeline specifically for this.
>For the encoder itself needs a parser engine and afaict the most popular is Jinja2 (C++, Python, relevant SO discussion).
>And of course this also implies that we need a virtual env setup for Python, which needs a separate discussion.

If you’re starting to talk about using continuous integration to test a generator for classes with one required constructor and one required member function, you know there’s something seriously wrong.  You’d need a whole pile of glue around it to create a subtarget where the generated device class is linked and instantiated.

Even then, CI doesn’t actually prevent rot, it just alerts you to it faster.  This proposal just increases the amount of required maintenance, as the continuous integration setup itself would need to be maintained.  There’s no justification for putting this much effort into scaffolding to support a tool that generates a tiny amount of boilerplate.

>Consider this a PoC, wanna see the C++ MAME alternative? Here:

You may not have noticed, but `nltool` does far more than generate empty boilerplate classes…

>Copyright holder is fixed with Couriersud as author, while I respect the work on netlist here (which is also not my domain) 1356 lines looks a bit too much for classifying it as a "simple" tool lol;

That’s because what it’s doing is non-trivial – it’s:
*A static compiler/optimiser for the matrices that are solved to simulate the netlists.  That is, it statically compiles netlists to C++ code to solve the specific matrices for those particular netlists.
* A tool for running netlist simulation with input stimuli supplied from a CSV file.
* An automatic declaration generator – it takes the netlist device source files and generates the headers containing the device declarations for inclusion into netlist sources.

You could possibly make a case that the code should be structured differently, but it’s not an unreasonable amount of code for what it does.  Also, the “simple” comment predates the addition of the static netlist compilation functionality – it was a lot simpler before that was added.

The copyright holder you’re complaining about is listed in the output of the help command, not included in generated sources.  Couriersud is, in fact, the copyright holder of `nltool` itself.

>There are Russian strings at the very end under `#if 0`;

So he added some test code for seeing how the standard library handles Unicode I/O on various platforms and commented it out rather than remove it.  He definitely isn’t the first person to do that, and he’s done a lot better than all the people who’ve added test code forgotten to even comment it out before pushing.

Given how far off the mark and borderline insulting your comments on nltool are, I’m honestly wondering whether you’re just trolling at this point.

**Comment by angelosa** [comment]
I'm not trolling, but seriously didn't expect this TL;DR about mirror climbing your point.

1. Apparently in your world everyone is a C++ developer virtuoso and everyone uses your environment to develop for MAME.  Good for you.
2. You dodged my point about Jinja2 and virtual env for Python entirely. You're definitely scared cause not your turf and e.g. prefer using SQLite in minimaws. That entirerly classifies as pet project too, see below.
3. The `#if 0` is my talk about not following convention about using `#ifdef UNUSED_FUNCTION` instead. And my whole point is that generating that code would require half the lines in Python.

Bottom line: given all this unnecessary antagonism and no concrete criticism over a code generator I'm left on the only option to strip any reference to `src\mame\etc\template_*` and move as my pet project.

**Comment by cuavas** [comment]
>Apparently in your world everyone is a C++ developer virtuoso and everyone uses your environment to develop for MAME. Good for you.

If you think writing a class with two standard members requires “a C++ developer virtuoso”, I really don’t know what to say.  Writing the empty class is trivially easy in any environment, all the way down to a plain text editor.

>The `#if 0` is my talk about not following convention about using `#ifdef UNUSED_FUNCTION` instead. And my whole point is that generating that code would require half the lines in Python.

Even if that’s the case, it’s still way off the mark:
* It’s obvious the netlist library has its own conventions, and attempts to be semi-independent of MAME even if MAME is its only major application.  There are lots of things that it doesn’t do the MAME way, which has been a source of contention on a few occasions.
* `#ifdef UNUSED_FUNCTION` was supposed to be used for disabling entire functions that aren’t used to suppress unused function warnings from the compiler.  The idea was to have a separate build configuration with the warning-as-error for unused functions disabled and the `UNUSED_FUNCTION` macro defined to stop these functions from rotting.  That never happened, so the functions rotted anyway.  But now that we’re on C++17, you should use the `[[maybe_unused]]` attribute instead – then the code will be checked by the compiler on every build.  The `#if UNUSED_FUNCTION` is no longer necessary and should be phased out now that there’s a proper C++ language facility.
* `#ifdef UNUSED_FUNCTION` was never supposed to be for disabling arbitrary pieces of code or things that change behaviour.  We’ve always used `#if 0` of stuff like that – there are over 600 examples of it in src/mame right now.

>You dodged my point about Jinja2 and virtual env for Python entirely. You're definitely scared cause not your turf and e.g. prefer using SQLite in minimaws. That entirerly classifies as pet project too, see below.

What do you even mean by “virtual env for Python”?  You think we should ship a virtual machine for running Python or something?  Why are you throwing random things around like SQLite?  We’ve been through this before:
* All of MAME’s Python scripts intentionally have no dependencies beyond the Python standard library so they work out-of-the-box for as many people as possible.  Any additional dependencies will make getting set up more difficult.
* SQLite3 isn’t a report generation language or template engine.  You couldn’t replace it with Jinja2 because it doesn’t even remotely solve the same class of problems – you’d still need to use some kind of database no matter what report generation framework you use.  I used `string.Template` for basic template functionality because it’s in the Python standard library, and hence doesn’t add dependencies.
* I originally made minimaws because there was a real demand from front-end authors for sample code to show how to do things like slot configuration based on `-listxml` output.  I’ve got very positive feedback from front-end authors.  The ROM identification features were added in response to seeing that dumping groups were using horribly outdated web-based ROM identification sites.  Once again, it was to address a real problem.

>Bottom line: given all this unnecessary antagonism and no concrete criticism over a code generator

My criticism is that it doesn’t significantly reduce effort – an empty device class is trivially easy to write – and something that *does* automate the more repetitive parts of writing devices would need a whole code refactoring engine to support it.

**Comment by angelosa** [comment]
Not sure what's your goal about continuing this conversation. I'll try to be brief because looks like you have no idea about a couple key points:

1. The whole idea is not about how trivial creating a device is. **It's the necessity of creating 2 files and changing 3 scripts and remembering all the conventions**. Five steps, do the math.
2. Out of sheer curiosity I've default [pylint-ed](https://pylint.pycqa.org/en/latest/) your `scripts\minimaws\lib` (pylint=2.11.1, astroid=2.8.4, python=3.7). _4.43_ rating, including some bad errors. And that doesn't cover the inherently obtuse bad design/reinventing the wheel by ignoring the vast [Python Software Index](https://pypi.org/) -> the correlated [virtual env](https://docs.python.org/3/tutorial/venv.html) necessary in covering it for reproducibility/testing/adding dependencies for even the basic tasks, cfr. swapping SQLite with something even remotely usable in an actual website or understanding what's the culprit about the dire startup performance (lacks [bulk insert](https://www.sqlite.org/lang_insert.html) my gut instinct?) or use Jinja2 for actual serious template engine **in a separate venv package than minimaws**.

**Comment by rb6502** [comment]
I've written a lot more devices than most contributors and I don't consider the boilerplate aspect to be difficult or tedious, especially compared to the actual implementation work.  Especially since the boilerplate varies wildly by what you're inheriting from, so no tool could actually cover all of the possibilities (every devices/bus/ subfolder has different requirements, for instance).

**Comment by cuavas** [comment]
>The whole idea is not about how trivial creating a device is. It's the necessity of creating 2 files and changing 3 scripts and remembering all the conventions. Five steps, do the math.

The fiddly part of that is adding the files to the build scripts, which your Python script didn’t even attempt to do.

In order to do actually do that, you’d need a proper Lua parser that can build an AST for the build scripts and inject the new stuff at an appropriate place.  This will need to change to a proper CMake script parser when we go there.  That’s a lot of work, and I see no evidence you’re prepared to actually do it.  Sucking in dependencies to do this isn’t going to fly – it’s not an important enough for that.

Also, you’d need a wizard-like UI for entering the necessary information, like the name of the variable, the headers to trigger inclusion, and so on, and then some UI for confirming the final changes to the build scripts (position of the insertions, etc.).  A question/answer interface in text mode isn’t adequate for this kind of thing.  It ends up being more effort than trying to do it by hand.

I can’t see how the effort to do this properly and then maintain it could possibly be worthwhile.

>Out of sheer curiosity I've default pylint-ed your scripts\minimaws\lib (pylint=2.11.1, astroid=2.8.4, python=3.7). 4.43 rating, including some bad errors. And that doesn't cover the inherently obtuse bad design/reinventing the wheel by ignoring the vast Python Software Index -> the correlated virtual env necessary in covering it for reproducibility/testing/adding dependencies for even the basic tasks, cfr. swapping SQLite with something even remotely usable in an actual website or understanding what's the culprit about the dire startup performance (lacks bulk insert my gut instinct?) or use Jinja2 for actual serious template engine in a separate venv package than minimaws.

This doesn’t make sense at all.

First of all, the primary purpose of minimaws is to serve as sample code for front-end authors.  It demonstrates a number of important tasks, including:
* Doing an object-relational mapping on MAME’s `-listxml` output and software lists.
* Slot configuration purely using static data, including slots that add more slots.
* Getting the overall emulation status flags for a system configuration.
* Finding software lists available for a given system configuration.
* Going in the opposite direction, finding a system or slot card compatible with a software list.

Any additional dependencies beyond the standard library would make it harder to get started with, and hence less accessible.  That’s the opposite of what you want for sample code.

SQLite is perfectly usable in an actual web site, particularly if most of the accesses are read-only.  Plenty of actual web sites (including the [SQLite site](https://sqlite.org/index.html) itself are powered by SQLite, and I’m successfully running a minimaws instance mainly to make a reasonably up-to-date [ROM/CHD identification service](https://arcade.vastheman.com/minimaws/romident) readily available.

You’re confusing initial database population and indexing with startup performance.  You obviously don’t load and index the database every time you serve a request.  That would be silly.  The startup performance and query performance are fine – I’ve ensured all the tables are indexed properly and none of the queries fall back to a table scan in their query plans.

Populating and indexing the database in the first place is fairly time-consuming, but it doesn’t need to happen very frequently, so it’s not a major issue.  I do take some steps to reduce overhead.  For example, I add 1024 machines per transaction to avoid waiting for filesystem sync unnecessarily (the number is arbitrary – making it too high makes the actual commit operations take longer due to the size of the rollback log).  I also create most of the indexes after inserting all the data as it’s more efficient to do that way than the have them updated for each row that’s inserted.

Sure, there are optimisation opportunities there, but it’s not really worth it for a relatively infrequent operation.  Throwing around buzzwords like “bulk insert” and assuming I have no clue what I’m doing is the wrong starting point.

**Comment by angelosa** [comment]
What is so hard to understand that it basically **fails basic linting with errors**?
Go buy a Python book (cit.) instead of wasting mine and your time please.

**Comment by cuavas** [comment]
>What is so hard to understand that it basically fails basic linting with errors?
>Go buy a Python book (cit.) instead of wasting mine and your time please.

Yet it actually works, runs correctly under both Python 2.7 and Python 3.4 or later, performs well enough in a WSGI container, and hasn’t resulted in me being pwned like what happened to MAWS itself.  But you keep moving the goalposts.  You only brought up minimaws as a kind of non sequitur, like the Unicode I/O test code with Russian strings in nltool.

All of this seems to be deflecting from the fact that the code in this PR doesn’t actually address the annoying or non-trivial parts of adding a device to MAME.
